crates/doc/src/inference.rs

```diff
@@ -1484,6 +1561,152 @@ impl Shape {
+    where
+        N: AsNode,
+    {
+        match node.as_node() {
+            crate::Node::Object(fields) => {
+                // See comment in `ObjShape::widen` about when to set `is_required`
+                // on newly encountered fields. Detects whether this is the
+                // very first time this location has seen an OBJECT shape.
+                let is_first_time = !self.type_.overlaps(types::OBJECT);
+                self.type_ = self.type_ | types::OBJECT;
+
+                if let None = self.object.additional {
+                    self.object.additional = Some(Box::new(Shape {
+                        type_: types::INVALID,
+                        provenance: Provenance::Inline,
+                        ..Default::default()
+                    }))
+                }
+
+                let hint = self.object.widen::<N>(fields, loc, is_first_time);
+
+                match (hint, loc) {
+                    (true, _) => true,
+                    (false, Location::Root) => self.object.properties.len() > MAX_ROOT_FIELDS,
+                    (false, _) => self.object.properties.len() > MAX_NESTED_FIELDS,
+                }
+            }
+
+            crate::Node::Array(arr) => {
+                let mut shape = self.array.additional.take().unwrap_or(Box::new(Shape {
+                    // Start out maximally constrained
+                    type_: types::INVALID,
+                    provenance: Provenance::Inline,
+                    ..Default::default()
+                }));
+
+                // Look at each element in the observed array and widen the shape to accept it
+                let hint = arr.iter().enumerate().fold(false, |accum, (idx, node)| {
+                    accum || shape.widen(node, loc.push_item(idx))
+                });
+
+                self.array.additional = Some(shape);
+                self.type_ = self.type_ | types::ARRAY;
+
+                hint
+            }
+            crate::Node::Bool(_) => {
+                self.type_ = self.type_ | types::BOOLEAN;
+                false
+            }
+            crate::Node::Bytes(_) => {
+                self.type_ = self.type_ | types::STRING;
+
+                let partial_stringshape = StringShape {
+                    content_encoding: Some("base64".to_string()),
+                    ..Default::default()
+                };
+
+                self.string = StringShape::union(self.string.clone(), partial_stringshape);
+                false
+            }
+            crate::Node::Null => {
+                self.type_ = self.type_ | types::NULL;
+                false
+            }
+            crate::Node::Number(_num) => {
+                self.type_ = self.type_ | types::INT_OR_FRAC;
+                false
+            }
+            crate::Node::String(_s) => {
+                self.type_ = self.type_ | types::STRING;
+                let partial_stringshape = StringShape {
+                    // TODO: implement format widening
```

Comment:
this needs doing to before we release for real. see [infer_string_shape](https://github.com/estuary/flow/blob/108864c33849e689da48a81f8ff7b65e02372f08/crates/schema-inference/src/inference.rs#L23C5-L37).

Though I am concerned about the performance aspect of these checks and we need to be careful.
Perhaps we always match over format on the very first string, and thereafter we match on format if and only if the LHS format is still set. Put differently, as soon as LHS `format` is unconstrained we stop checking the RHS.

We need test on this also -- perhaps ripped from `schema-inference` crate.


Also update min_length / max_length?

Reply:
This makes sense. I ended up constraining it a little bit more: only first-time strings get the full inference, then subsequent ones get validated against the format and if they don't match, lose the format.

Reply:
No, because there can be more nuanced handling for `format` in StringShape::union.

One current example is widening `format: integer` => `format: number`.

Reply:
Ah yep, good point. Changed and tested 👍